driver-core/src/main/java/com/datastax/driver/core/ControlConnection.java

```diff
@@ -258,11 +257,14 @@ private Connection tryConnect(Host host, boolean isInitialConnection) throws Con
             // the node we're connecting to.
             refreshNodeListAndTokenMap(connection, cluster, isInitialConnection, true);
 
-            // Note that refreshing the schema will trigger refreshNodeListAndTokenMap since table == null
-            // We want that because the token map was not properly initialized by the first call above, since it requires the list of keyspaces
-            // to be loaded.
             logger.debug("[Control connection] Refreshing schema");
-            refreshSchema(connection, null, null, cluster, isInitialConnection);
+            refreshSchema(connection, null, null, cluster);
+
+            // We need to refresh the node list again;
+            // We want that because the token map was not properly initialized by the first call above,
+            // since it requires the list of keyspaces to be loaded.
+            refreshNodeListAndTokenMap(connection, cluster, false, true);
```

Comment:
This was called from inside refreshSchema in certain situations; that had to be pulled out of that method; when initiating the connection this method was effectively always called, so we simply call it here (synchronously because we need the information ready when the method returns).